src/rnnlm/arpa-sampling.h

```diff
@@ -0,0 +1,161 @@
+enum {
+  kEps = 0,
+  // kDisambig,
+  kBos, kEos, kUnk
+};
+
+typedef std::vector<int32> HistType;
+typedef unordered_map<int32, std::pair<BaseFloat, BaseFloat> > WordToProbsMap; 
+typedef unordered_map<HistType, WordToProbsMap, IntVectorHasher> NgramType;
+typedef unordered_map<HistType, BaseFloat, IntVectorHasher> HistWeightsType;
+
+class ArpaSampling : public ArpaFileParser {
+ public:
+  // constructor
+  explicit ArpaSampling(ArpaParseOptions options, fst::SymbolTable* symbols)
+     : ArpaFileParser(options, symbols) { 
+       ngram_order_ = 0;
+       num_words_ = 0;
+       bos_symbol_ = "<s>";
+       eos_symbol_ = "</s>";
+       unk_symbol_ = "<unk>";
+  }
+  // Compute the probability of a given sentence with ngram_order LM
+  BaseFloat ComputeSentenceProb(const std::vector<int32>& test_sentence);
+  
+  // Test the read-in model by computing probs of all sentences with ngram_order LM
+  BaseFloat ComputeAllSentencesProb(const std::vector<std::vector<int32> >& test_sentences);
+  
+  void TestReadingModel();
+
+  void TestProbs(std::istream &is, bool binary);
+
+  void TestPdfsEqual();
+
+  // print history
+  void PrintHist(const HistType& h);
+  
+  void ReadHistories(std::istream &is, bool binary);
+
+  void ReadSentences(std::istream &is, std::vector<std::vector<int32> >* sentences);
+  
+ protected:
+  // ArpaFileParser overrides.
+  virtual void HeaderAvailable(); 
+  virtual void ConsumeNGram(const NGram& ngram);
+  virtual void ReadComplete() {}
+
+ private:
+  // This function returns the log probability of a ngram term from the ARPA LM
+  // if it is found; it backoffs to the lower order model when the ngram term 
+  // does not exist.
+  BaseFloat GetProb(int32 order, int32 word, const HistType& history);
+
+  // Get the back-off weight of a ngram in the read-in model
+  BaseFloat GetBackoffWeight(int32 order, int32 word, const HistType& history);
+  
+  // For test: randomly generate histories
+  void RandomGenerateHistories();
+
+  // Compute a pdf of words in the vocab given a history
+  void ComputeWordPdf(const HistType& history, std::vector<std::pair<int32, BaseFloat> >* pdf);
+  
+  // Compute weights of given histories
+  void ComputeHistoriesWeights();
+
+  // Compute weighted pdf given all histories
+  void ComputeWeightedPdf(std::vector<std::pair<int32, BaseFloat> >* weighted_pdf);
+  
+  // N-gram order of the read-in LM.
+  int32 ngram_order_;
+  
+  // num_words
+  int32 num_words_;
+
+  // Bos symbol
+  std::string bos_symbol_;
+
+  // Eos symbol
+  std::string eos_symbol_;
+
+  // Unk symbol
+  std::string unk_symbol_;
+
+  // Vocab
+  std::vector<std::pair<std::string, int32> > vocab_;
+
+  // Counts of each ngram
+  std::vector<int32> ngram_counts_;
+
+  // N-gram probabilities.
+  std::vector<NgramType> probs_;
+
+  // Histories' weights
+  HistWeightsType hists_weights_;
+  
+  // The given N Histories
+  std::vector<HistType> histories_;
```

Comment:
This class should only store information about the ngram model (read from the arpa file). Histories should just be a paramter you pass in order to get the prob-distributions.